nixos/modules/services/infrastructure/terraform-cloud-agent/default.nix

```
@@ -0,0 +1,144 @@
+      };
+
+      log-level = lib.mkOption {
+        type = lib.types.nullOr (lib.types.enum [ "trace" "debug" "info" "warn" "error" ]);
+        default = null;
+        description = lib.mkDoc "Level at which log messages should be emitted.";
+        example = "warn";
+      };
+
+      address = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+        description = lib.mkDoc "HTTP or HTTPS address of the Terraform Cloud API.";
+        example = "https://app.terraform.io";
+      };
+
+      accept = lib.mkOption {
+        type = lib.types.commas;
+        default = "plan,apply,policy,assessment";
+        description = lib.mkDoc "Optional string of comma-separated job types that this agent may run. Acceptable job types are 'plan', 'apply', 'policy', and 'assessment'. Do not put whitespace in between entries."
+        example = "plan,apply";
+      };
+
+      auto-update = lib.mkOption {
+        type = lib.types.str;
+        readOnly = true;
+        default = "disabled";
+        description = lib.mkDoc "Controls automatic core updates behavior. Not acceptable in NixOS, so disabled."
+      };
+    };
+  };
+
+  config = {
+    systemd.services.terraform-cloud-agent = lib.mkIf cfg.enable {
+      description = "Agent that executes plans from Terraform Cloud";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+
+      environment = {
+        TFC_AGENT_DATA_DIR = "%S";
+        TFC_AGENT_CACHE_DIR = "%C";
+      };
+
+      script = ''
+        export TFC_AGENT_TOKEN=$(cat ${cfg.tokenPath})
+    
+        ${lib.getExe pkgs.terraform-cloud-agent} ${pkgs.toGNUCommandLineShell { mkOptionName = k: "-${k}"; } cfg.flags}
+      '';
+
+      serviceConfig = {
+        StateDirectory = "terraform-cloud-agent";
+        CacheDirectory = "terraform-cloud-agent";
+
+        ConfigurationDirectory = "terraform-cloud-agent";
+
+        # User and group
+        User = cfg.user;
+        Group = cfg.group;
+
+        # Below shamelessly cribbed from the nginx module: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/nginx/default.nix
+
+        # Security
+        NoNewPrivileges = true;
+
+        # Sandboxing (sorted by occurrence in https://www.freedesktop.org/software/systemd/man/systemd.exec.html)
+        ProtectSystem = "strict";
+        ProtectHome = mkDefault true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        ProtectHostname = true;
+        ProtectClock = true;
+        ProtectKernelTunables = true;
+        ProtectKernelModules = true;
+        ProtectKernelLogs = true;
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        LockPersonality = true;
+        MemoryDenyWriteExecute = !((builtins.any (mod: (mod.allowMemoryWriteExecute or false)) cfg.package.modules) || (cfg.package == pkgs.openresty));
```

These options don't exist for tfc-agent, right?

```suggestion
        MemoryDenyWriteExecute = true
```

---

Lol, yeah, good catch, thanks.